caffe2/onnx/onnx_exporter.cc

```diff
@@ -933,14 +933,17 @@ ConvertedResult OnnxExporter::CreateGemmNodes(
     axis = it->second->i();
   }
   if (x_shape.dims().size() > 2) {
+    std::vector<AttributeProto> xattrs = {MakeAttribute("shape",
```

Comment:
Why do we need this attribute? We are supposed to read the shape from second input.

Reply:
@yinghai thanks for pointing this out. Here is the problem: onnx-tensorrt requires this attribute for opset versions less than 5. When I declare version 5+, it fails with recent ONNX models having Reshape op weights passed as floats. To resolve this, I found this solution (it requires code change in onnx-tensorrt):
```
DEFINE_BUILTIN_OP_IMPORTER(Reshape) {
  auto input = inputs.at(0);
  nvinfer1::Dims new_shape;
  if( ctx->getOpsetVersion() >= 5 ) {
    ASSERT(inputs.size() == 2, ErrorCode::kINVALID_NODE);
    auto new_shape_input = inputs.at(1);
    ASSERT(new_shape_input.is_weights(), ErrorCode::kUNSUPPORTED_NODE);
    ShapedWeights new_shape_weights = new_shape_input.weights();
    ASSERT(new_shape_weights.shape.nbDims == 1, ErrorCode::kINVALID_NODE);
    if(new_shape_weights.type == ::ONNX_NAMESPACE::TensorProto::FLOAT) {
      float const* new_shape_ptr =
          static_cast<float const*>(new_shape_weights.values);
      new_shape.nbDims = new_shape_weights.shape.d[0];
      for(int i = 0; i < new_shape.nbDims; ++i) {
        new_shape.d[i] = std::lround(new_shape_ptr[i]);
      }
    } else {
      ASSERT(new_shape_weights.type == ::ONNX_NAMESPACE::TensorProto::INT64,
          ErrorCode::kINVALID_NODE);
      int64_t const* new_shape_ptr =
          static_cast<int64_t const*>(new_shape_weights.values);
      new_shape.nbDims = new_shape_weights.shape.d[0];
      std::copy(new_shape_ptr, new_shape_ptr + new_shape.nbDims, new_shape.d);
    }
  } else {
    OnnxAttrs attrs(node);
    new_shape = attrs.get<nvinfer1::Dims>("shape");
  }
...etc...

```

But I'm not quite sure that it's the right thing to do. @benbarsdell ?

Reply:
> When I declare version 5+, it fails with recent ONNX models having Reshape op weights passed as floats.

I think we need to fix the part where we always pass weights as floats:
https://github.com/pytorch/pytorch/blob/1e73ab25f58b5afe1b10dbb65ab30c6251a66e35/caffe2/contrib/tensorrt/tensorrt_tranformer.cc#L301

Replace the line above with this piece of code might just solve it: 
https://github.com/pytorch/pytorch/blob/1e73ab25f58b5afe1b10dbb65ab30c6251a66e35/caffe2/opt/onnxifi_transformer.cc#L227-L240

Reply:
Done